resources/sparql.js.bnf

```diff
@@ -0,0 +1,251 @@
+SubstringExpression   ::=   <'SUBSTR'> <'('> Expression <','> Expression ( <','> Expression )? <')'>
+StrReplaceExpression    ::=   <'REPLACE'> <'('> Expression <','> Expression <','> Expression ( <','> Expression )? <')'>
+ExistsFunc    ::=   <'EXISTS'> GroupGraphPattern
+NotExistsFunc   ::=   <'NOT'> WS <'EXISTS'> GroupGraphPattern
+Aggregate   ::=     'COUNT' WS <'('> WS 'DISTINCT'? WS ( '*' | Expression ) WS <')'> WS
+| 'SUM' WS <'('> WS 'DISTINCT'? Expression <')'>
+| 'MIN' <'('>  WS 'DISTINCT'? Expression <')'>
+| 'MAX' <'('>  WS 'DISTINCT'? Expression <')'>
+| 'AVG' <'('>  WS 'DISTINCT'? Expression <')'>
+| 'SAMPLE' <'('>  WS 'DISTINCT'? Expression? Expression <')'>
+| 'GROUP_CONCAT' <'('> WS 'DISTINCT'? Expression ( <';'> WS 'SEPARATOR' WS <'='> WS String WS )? <')'>
+iriOrFunction	  ::=  	iri ArgList?
+RDFLiteral	  ::=  	String WS ( LANGTAG | ( '^^' iri ) )? WS
+NumericLiteral	  ::=  	NumericLiteralUnsigned WS | NumericLiteralPositive WS | NumericLiteralNegative WS
+<NumericLiteralUnsigned>	  ::=  	INTEGER | DECIMAL | DOUBLE
+<NumericLiteralPositive>	  ::=  	INTEGER_POSITIVE | DECIMAL_POSITIVE | DOUBLE_POSITIVE
+<NumericLiteralNegative>	  ::=  	INTEGER_NEGATIVE | DECIMAL_NEGATIVE | DOUBLE_NEGATIVE
+BooleanLiteral	  ::=  	'true' WS | 'false' WS
+<String>	  ::=  	STRING_LITERAL1 | STRING_LITERAL2 | STRING_LITERAL_LONG1 | STRING_LITERAL_LONG2
+iri	  ::=  ( IRIREF | PrefixedName ) WS
+PrefixedName	  ::=  	( PNAME_LN WS ) | ( PNAME_NS WS)
+BlankNode	  ::=  	BLANK_NODE_LABEL | ANON
+PrettyPrint ::= <'PRETTY-PRINT'> WS
+IRIREF	  ::=  	#"<[^<>\"{}|^`\x00-\x20]*>" WS
+<PNAME_NS>	  ::=  	PN_PREFIX? ':'
+<PNAME_LN>	  ::=  	PNAME_NS PN_LOCAL '*'?
+BLANK_NODE_LABEL	  ::=  	'_:' ( PN_CHARS_U | #"[0-9]" ) ((PN_CHARS|'.')* PN_CHARS)?
+<VAR1>	  ::=  	<'?'> VARNAME
+<VAR2>	  ::=  	<'$'> VARNAME
+LANGTAG ::= #"@[a-zA-Z]+(-[a-zA-Z0-9]+)*" WS
+<INTEGER>	  ::=  	#"[0-9]+"
+<DECIMAL>	  ::=  #"[0-9]*\.[0-9]*"
+<DOUBLE>	  ::=  	#"[0-9]+\.[0-9]*|(\.[0-9]+)|([0-9]+)" EXPONENT
+<INTEGER_POSITIVE>	  ::=  	'+' INTEGER
+<DECIMAL_POSITIVE>	  ::=  	'+' DECIMAL
+<DOUBLE_POSITIVE>	  ::=  	'+' DOUBLE
+<INTEGER_NEGATIVE>	  ::=  	'-' INTEGER
+<DECIMAL_NEGATIVE>	  ::=  	'-' DECIMAL
+<DOUBLE_NEGATIVE>	  ::=  	'-' DOUBLE
+EXPONENT	  ::=  	#"[eE][+-]?[0-9]+"
+<STRING_LITERAL1>	  ::=  	<"'"> ( #"[^\x27\x5C\x0A\x0D]" | ECHAR )* <"'">
+<STRING_LITERAL2>	  ::=  	<'"'> ( #"[^\x27\x5C\x0A\x0D]" | ECHAR )* <'"'>
+<STRING_LITERAL_LONG1>	  ::=  	"'''" ( ( "'" | "''" )? ( #"[^'\\]" | ECHAR ) )* "'''"
+<STRING_LITERAL_LONG2>	  ::=  	'"""' ( ( '"' | '""' )? ( #"[^'\\]" | ECHAR ) )* '"""'
+ECHAR	  ::=  	#"\\[tbnrf]"
+NIL	  ::=  	'(' WS* ')'
+<WS>	  ::=  	<#"[\x20\x09\x0D\x0A]*\#[^\n]*\n*[\x20\x09\x0D\x0A]*|[\x20\x09\x0D\x0A]*">
+ANON	  ::=  	'[' WS* ']'
+<PN_CHARS_BASE>	  ::=  	#"(?u)[a-zA-Z]|[\u00C0-\u00D6]|[\u00D8-\u00F6]|[\u00F8-\u02FF]|[\u0370-\u037D]|[\u037F-\u1FFF]|[\u200C-\u200D]|[\u2070-\u218F]|[\u2C00-\u2FEF]|[\u3001-\uD7FF]|[\uF900-\uFDCF]|[\uFDF0-\uFFFD]|[\u{10000}-\u{EFFFF}]"
```

Comment:
This is the line that required different syntax. The important bits are the flag prefix: (?u) and the unicode character syntax for multibyte codepoints: [\u{10000}-\u{EFFFF}]

Reply:
Is there a way to reuse most of the file so we don't have to make edits in two places going forward?

For example, could we define a `sparql.common.bnf` file that contains everything except for line 241 of "sparql.bnf" and then define one `sparql.pn_chars_base.js.bnf` and a `sparql.pn_chars_base.jvm.bnf` that contain only line 241 with the respective syntax for the platform, and then have helper functions that loads the right resources for each platform and concatenates them into a single grammar?

That's just a thought. I would like to avoid needing to edit two different grammars whenever we need to make a change going forward if possible.

Reply:
Good thought, I've separated out the offending line, should be much more maintainable this way.